File: Python/Maths/Happy Number Algorithm.py
def is_happy(n):
    seen = set()

    while n != 1:
        current = n
        sum_of_squares = 0

        while current != 0:
            digit = current % 10
            sum_of_squares += digit * digit
            current //= 10

        if sum_of_squares in seen:
            return False  # Entered a cycle containing 4, so it's an unhappy number

        seen.add(sum_of_squares)
        n = sum_of_squares

    return True  # Reached 1, so it's a happy number

number = int(input("Enter a number: "))
is_happy_number = is_happy(number)
print(f"{number} is a happy number: {is_happy_number}")

File: C/Maths/Happy Number Algorithm.c
#include <stdio.h>

int isHappy(int n) {
    int seen[1000] = {0};

    while (n != 1) {
        int current = n;
        int sum_of_squares = 0;

        while (current != 0) {
            int digit = current % 10;
            sum_of_squares += digit * digit;
            current /= 10;
        }

        if (seen[sum_of_squares] == 1) {
            return 0; // Entered a cycle containing 4, so it's an unhappy number
        }

        seen[sum_of_squares] = 1;
        n = sum_of_squares;
    }

    return 1; // Reached 1, so it's a happy number
}

int main() {
    int number;
    printf("Enter a number: ");
    scanf("%d", &number);
    int isHappyNumber = isHappy(number);
    printf("%d is a happy number: %s\n", number, isHappyNumber ? "true" : "false");
    return 0;
}
